### PR TITLE
feat: collection level permissions

### PR DESCRIPTION
**Pull request by TorstenDittmann**
- implements `documentSecurity` to collections
- respect `documentSecurity` on all CRUD operations

### CODE REVIEW

File: src/Database/Database.php [hunk 1]
```diff
@@ -3558,6 +3604,13 @@ public function find(string $collection, array $queries = [], ?int $timeout = nu
 
         $collection = $this->silent(fn () => $this->getCollection($collection));
 
+        if ($collection->getAttribute('documentSecurity', true) === false) {
+            $authorization = new Authorization(self::PERMISSION_READ);
+            if ($authorization->isValid($collection->getRead())) {
+                $skipAuth = true;
+            }
+        }
```
Comment (stnguyen90) [inline src/Database/Database.php:3612]: I don't think this check should be only for `documentSecurity == true`. What if `documentSecurity` is `true` AND there are collection-level permissions?

Same for the `count()` method.

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: src/Database/Database.php [hunk 2]
```diff
@@ -2090,9 +2113,11 @@ public function getDocument(string $collection, string $id, array $queries = [])
         if ($cache = $this->cache->load($cacheKey, self::TTL)) {
             $document = new Document($cache);
 
-            if ($collection->getId() !== self::METADATA
-                && !$validator->isValid($document->getRead())) {
-                return new Document();
+            if ($collection->getId() !== self::METADATA) {
+                $isValid = $documentSecurity ? $validator->isValid($document->getRead()) : $validator->isValid($collection->getRead());
```
Comment (stnguyen90) [inline src/Database/Database.php:2117]: Would this ignore collection level permissions if `documentSecurity` was `true`?

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: src/Database/Database.php [hunk 3]
```diff
@@ -3809,10 +3905,18 @@ public function count(string $collection, array $queries = [], ?int $max = null)
             throw new Exception("Collection not found");
         }
 
+        if ($collection->getAttribute('documentSecurity', true) === false) {
+            $authorization = new Authorization(self::PERMISSION_READ);
+            if ($authorization->isValid($collection->getRead())) {
+                $skipAuth = true;
+            }
+        }
```
Comment (stnguyen90) [inline src/Database/Database.php:3913]: The outer if check should be removed here, right?

Comment (TorstenDittmann) [reply src/Database/Database.php:3913]: good catch, leftover

File: src/Database/Database.php [hunk 4]
```diff
@@ -3455,21 +3542,21 @@ private function deleteCascade(Document $collection, Document $relatedCollection
...
             case Database::RELATION_ONE_TO_ONE:
                 $this->skipRelationships(fn () =>
-                    $this->deleteDocument(
-                        $relatedCollection->getId(),
-                        $value->getId()
-                    ));
+                $this->deleteDocument(
+                    $relatedCollection->getId(),
+                    $value->getId()
+                ));
```
Comment (stnguyen90) [inline src/Database/Database.php:3548]: I think this needs to check if `$value == null` before deleting or else you we get:

image

File: src/Database/Database.php [hunk 5]
```diff
@@ -3455,21 +3542,21 @@ private function deleteCascade(Document $collection, Document $relatedCollection
...
                 foreach ($value as $relation) {
                     $this->skipRelationships(fn () =>
-                        $this->deleteDocument(
-                            $relatedCollection->getId(),
-                            $relation->getId()
-                        ));
+                    $this->deleteDocument(
+                        $relatedCollection->getId(),
+                        $relation->getId()
+                    ));
```
Comment (stnguyen90) [inline src/Database/Database.php:3559]: Should this always skip? If it does, we'll have a limitation where related documents will only delete 1 level deep. For example, if i have:

> Quiz -> Questions -> Answer Choices

and I delete the Quiz, the Questions will be deleted, but the Answer Choices won't be.

**Comment by stnguyen90** [comment]
When setting a relationship, if a document with the specified ID doesn't exist, the API call will still succeed, but the relationship will be set to null.

File: src/Database/Database.php (hunk 5 above)
Comment (stnguyen90) [reply src/Database/Database.php]: I've created a task for us to track this.

File: src/Database/Database.php [hunk 6]
```diff
@@ -2410,6 +2470,13 @@ public function createDocument(string $collection, Document $document): Document
     {
         $collection = $this->silent(fn () => $this->getCollection($collection));
 
+        if ($collection->getId() !== self::METADATA) {
+            $authorization = new Authorization(self::PERMISSION_CREATE);
+            if (!$authorization->isValid($collection->getCreate())) {
```
[the remark this answers is not in the document]
Comment (stnguyen90) [reply src/Database/Database.php:2475]: @fogelito, I don't think we need to for `createDocument()`, right? That should only be on the collection level.

**Review by stnguyen90 — APPROVED** [review]

File: tests/Database/Base.php [hunk 7]
```diff
@@ -9966,6 +9971,663 @@ public function testManyToManyRelationshipKeyWithSymbols(): void
...
+     * @return array<Document>
+     */
+    public function testCollectionPermissionsCreateWorks(Document $collection): array
+    {
+        Authorization::cleanRoles();
+        Authorization::setRole(Role::users()->toString());
+
+        $document = static::getDatabase()->createDocument($collection->getId(), new Document([
+            '$id' => ID::unique(),
+            '$permissions' => [
```
Comment (fogelito) [inline tests/Database/Base.php:10015]: If documentSecurity is false for $collection , do we need to pass permissions on createDocument?

File: src/Database/Database.php [hunk 8]
```diff
@@ -651,6 +674,36 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+        }
+
+        $collection = $this->silent(fn () => $this->getCollection($id));
+
+        $collection
+            ->setAttribute('$permissions', $permissions)
+            ->setAttribute('documentSecurity', $documentSecurity);
+
+        $collection = $this->silent(fn () => $this->updateDocument(self::METADATA, $collection->getId(), $collection));
+
```
Comment (fogelito) [inline src/Database/Database.php:701]: Do we need to clean cache?

Comment (fogelito) [reply src/Database/Database.php:701]: Can you add a test for the method?

Comment (TorstenDittmann) [reply src/Database/Database.php:701]: This should be covered by `$this->updateDocument`. Adding a test 👍🏻

**Pull request merged**